(feat) Install the CSI manila on an existing MKS cluster (#130)
ovh/public-cloud-examples: containers-orchestration/managed-kubernetes/install-csi-manila/README.md

* A [Managed Kubernetes Service cluster](https://www.ovhcloud.com/en/public-cloud/kubernetes/) that is connected to a private network

## Configure the deployment

### Configure the Terraform OVH provider

Create an OVHcloud API token:

https://api.ovh.com/createToken?GET=/\*&POST=/\*&PUT=/\*&DELETE=/\*

Configure Terraform with this token:

```bash
vim ovhrc.sh
```

```bash
export OVH_ENDPOINT="ovh-eu"
export OVH_APPLICATION_KEY="<your_application_key>"
export OVH_APPLICATION_SECRET="<your_application_secret>"
export OVH_CONSUMER_KEY="<your_consumer_key>"
export OVH_CLOUD_PROJECT_SERVICE="<your_public_cloud_project_ID>"
```

## Customize the deployment

Configure Terraform with your OVH public cloud project id:

```bash
vim terraform.tfvars
```

```bash
service_name = "<your_public_cloud_project_id"
```

Configure Terraform with your Managed Kubernetes Cluster id:

```bash
vim terraform.tfvars
```

```bash
mks_cluster_id = "<your_mks_cluster_id>"
```

If your Managed Kubernetes Service cluster is in a 3AZ region, choose in which zone to create the share network (example: eu-west-par-a):

```bash
vim terraform.tfvars
```

```bash

> share_availability_zone = "<your_share_availability_zone>"

Reviewer: ```suggestion
```

> ```

Reviewer: ```suggestion
```

**Requirements:**

You need the following:
* [Terraform](https://www.terraform.io/) installed
* An [OVHcloud Public Cloud project](https://www.ovhcloud.com/en/public-cloud/)
* A [Managed Kubernetes Service cluster](https://www.ovhcloud.com/en/public-cloud/kubernetes/) that is connected to a private network

## Configure the deployment

### Configure the Terraform OVH provider

Create an OVHcloud API token:

https://api.ovh.com/createToken?GET=/\*&POST=/\*&PUT=/\*&DELETE=/\*

Configure Terraform with this token:

```bash
vim ovhrc.sh
```

```bash
export OVH_ENDPOINT="ovh-eu"
export OVH_APPLICATION_KEY="<your_application_key>"
export OVH_APPLICATION_SECRET="<your_application_secret>"
export OVH_CONSUMER_KEY="<your_consumer_key>"
export OVH_CLOUD_PROJECT_SERVICE="<your_public_cloud_project_ID>"
```

## Customize the deployment

Configure Terraform with your OVH public cloud project id:

```bash
vim terraform.tfvars
```

```bash
service_name = "<your_public_cloud_project_id"
```

Configure Terraform with your Managed Kubernetes Cluster id:

```bash
vim terraform.tfvars
```

```bash
mks_cluster_id = "<your_mks_cluster_id>"
```

> If your Managed Kubernetes Service cluster is in a 3AZ region, choose in which zone to create the share network (example: eu-west-par-a):

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

> ```bash

Reviewer: ```suggestion
```

> vim terraform.tfvars

Reviewer: ```suggestion
```

> ```

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

> ```bash

Reviewer: ```suggestion
```

# Install Manila CSI on a Managed Kubernetes Service (MKS) cluster

**Requirements:**

You need the following:
* [Terraform](https://www.terraform.io/) installed
* An [OVHcloud Public Cloud project](https://www.ovhcloud.com/en/public-cloud/)
* A [Managed Kubernetes Service cluster](https://www.ovhcloud.com/en/public-cloud/kubernetes/) that is connected to a private network

## Configure the deployment

### Configure the Terraform OVH provider

Create an OVHcloud API token:

https://api.ovh.com/createToken?GET=/\*&POST=/\*&PUT=/\*&DELETE=/\*

Configure Terraform with this token:

```bash
vim ovhrc.sh
```

```bash
export OVH_ENDPOINT="ovh-eu"
export OVH_APPLICATION_KEY="<your_application_key>"
export OVH_APPLICATION_SECRET="<your_application_secret>"
export OVH_CONSUMER_KEY="<your_consumer_key>"
export OVH_CLOUD_PROJECT_SERVICE="<your_public_cloud_project_ID>"
```

> 

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

> # Install Manila CSI on a Managed Kubernetes Sevice cluster

Reviewer: ```suggestion
# Install Manila CSI on a Managed Kubernetes Service (MKS) cluster
```

**Requirements:**

You need the following:
* [Terraform](https://www.terraform.io/) installed
* An [OVHcloud Public Cloud project](https://www.ovhcloud.com/en/public-cloud/)
* A [Managed Kubernetes Service cluster](https://www.ovhcloud.com/en/public-cloud/kubernetes/) that is connected to a private network

## Configure the deployment

### Configure the Terraform OVH provider

Create an OVHcloud API token:

https://api.ovh.com/createToken?GET=/\*&POST=/\*&PUT=/\*&DELETE=/\*

Configure Terraform with this token:

```bash
vim ovhrc.sh
```

```bash
export OVH_ENDPOINT="ovh-eu"
export OVH_APPLICATION_KEY="<your_application_key>"
export OVH_APPLICATION_SECRET="<your_application_secret>"
export OVH_CONSUMER_KEY="<your_consumer_key>"
export OVH_CLOUD_PROJECT_SERVICE="<your_public_cloud_project_ID>"
```

> ### Configure the Terraform OpenStack provider

Reviewer: Why does we need an openstack user?

> ### Configure the Terraform OpenStack provider

Reviewer: ```suggestion
```

> Use this documentation to create an OpenStack user:

Reviewer: ```suggestion
```

> https://docs.ovhcloud.com/en/guides/public-cloud/cross-functional/create-and-delete-a-user

Reviewer: ```suggestion
```

> And this documentation to download the openrc.sh file:

Reviewer: ```suggestion
```

> https://docs.ovhcloud.com/en/guides/public-cloud/cross-functional/compute-set-openstack-environment-variables

Reviewer: ```suggestion
```

## Customize the deployment

Configure Terraform with your OVH public cloud project id:

```bash
vim terraform.tfvars
```

```bash
service_name = "<your_public_cloud_project_id"
```

Configure Terraform with your Managed Kubernetes Cluster id:

```bash
vim terraform.tfvars
```

```bash
mks_cluster_id = "<your_mks_cluster_id>"
```

If your Managed Kubernetes Service cluster is in a 3AZ region, choose in which zone to create the share network (example: eu-west-par-a):

```bash
vim terraform.tfvars
```

```bash

> share_availability_zone = "<your_share_availability_zone>"

Reviewer: it seems that it is no more used with the new TF resource

```

## Deploy the stack

```bash
source ovhrc.sh

> source openrc.sh

Reviewer: ```suggestion
```

```

## Test

Create a Persistent Volume Claim:

```bash
kubectl apply -f test/pvc.yaml
```

Check the status of the PVC:
```bash

> kubectl get pvc nfs-share-pvc

Reviewer: ```suggestion
kubectl get pvc nfs-share-fs-pvc
```

> When the Manila share is created, create a deployment:

Reviewer: ```suggestion
Check the status of the PVC:
```bash
kubectl get pvc nfs-share-pvc
```

Check until the Manila share and the PVC is Bound:

```share
$ kubectl get pvc nfs-share-pvc
NAME            STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS     VOLUMEATTRIBUTESCLASS   AGE
nfs-share-pvc   Bound    pvc-db49f8e5-e1f7-42d3-8f5a-f1da9a69da0a   150Gi      RWX            csi-manila-nfs   <unset>                 2m16s
```

Create a deployment:
```

```bash

> kubectl apply -f deploy.yaml

Reviewer: ```suggestion
kubectl apply -f test/deploy.yaml
```

```

Verify that the pod is running:

```bash
kubectl get pod
```

Verify that you can scale the deployment (multi-attach volume):

```bash
kubectl scale deploy/nginx-deployment --replicas=2
kubectl get pod

> ```

Reviewer: ```suggestion
```

To verify RWX functionality, connect to one pod and create a file in the mounted directory (e.g., /usr/share/nginx/html). Then connect to the second pod and confirm the file is visible:

```bash
MY_POD=$(kubectl get po -o name | head -n 1)
echo $MY_POD
MY_POD_2=$(kubectl get po -o name | head -n 2)
echo $MY_POD_2

#Create a file in the pod number 1
kubectl exec $MY_POD -it -- touch /usr/share/nginx/html/index.html

#Display it in the pod number two
kubectl exec $MY_POD_2 -it -- ls
```

Your Manila share exposed through NFS is functioning correctly!
```
